### PR TITLE
refactor(hgraph): move duplicate tracking to graph layer

### PR DESCRIPTION
**Pull request by inabao**
## Summary
- move HGraph duplicate tracking from `LabelTable` into graph-level duplicate trackers and wire `support_duplicate` through HGraph and Pyramid graph params
- restore legacy HGraph duplicate index compatibility during deserialize/search, including legacy metadata mapping and duplicate payload recovery for flat and compressed graph storage
- add deterministic Pyramid duplicate tests that validate path-aware visibility semantics alongside the existing duplicate regression coverage

## Testing
- `make release -j4`
- `./build-release/tests/functests \"Pyramid Duplicate Path Semantics Same Path\"`
- `./build-release/tests/functests \"Pyramid Duplicate Path Semantics Prefix Descendant\"`
- `./build-release/tests/functests \"Pyramid Duplicate Path Semantics Shared Prefix Visibility\"`
- `./build-release/tests/functests \"Pyramid Duplicate Path Semantics Negative Control\"`
- `./build-release/tests/functests \"Pyramid Duplicate Test\"`
- `./build-release/tests/functests \"Pyramid Duplicate ID Test\"`

### CODE REVIEW

File: src/algorithm/hgraph.cpp [hunk 1]
```diff
@@ -1690,6 +1695,9 @@ HGraph::resize(uint64_t new_size) {
         pool_ = std::make_shared<VisitedListPool>(1, allocator_, new_size_power_2, allocator_);
         this->label_table_->Resize(new_size_power_2);
         bottom_graph_->Resize(new_size_power_2);
+        if (this->support_duplicate_ && bottom_graph_->GetDuplicateTracker()) {
+            bottom_graph_->GetDuplicateTracker()->Resize(new_size_power_2);
+        }
```
Comment (inabao) [inline src/algorithm/hgraph.cpp]: The resizing of the Duplicate Tracker should be handled within bottom_graph; no separate handling is required here.

File: src/algorithm/pyramid.cpp [hunk 2]
```diff
@@ -506,6 +514,7 @@ Pyramid::Add(const DatasetPtr& base, AddMode mode) {
         auto inner_id = static_cast<InnerIdType>(i + local_cur_element_count);
         const auto* vector = data_vectors + dim_ * data_bias;
         int no_build_level_index = 0;
+        int add_count = 0;
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/algorithm/pyramid.cpp]: Removed the unused `add_count` declaration in a follow-up commit.

File: examples/cpp/107_index_pyramid.cpp [hunk 3]
```diff
@@ -15,152 +14,85 @@
...
-            selected_levels.emplace_back(level2[distr(mt) % level2.size()]);
-        }
-        if (num_levels == 3) {
-            selected_levels.emplace_back(level3[distr(mt) % level3.size()]);
+    // Generate base data
+    for (int i = 0; i < NUM; ++i) {
+        ids[i] = i;
+        paths[i] = "a/b/c";
+        for (int j = 0; j < DIM; ++j) {
+            vectors[i * DIM + j] = static_cast<float>(rand()) / static_cast<float>(RAND_MAX);
```
[the remark this answers is not in the document]
Comment (inabao) [reply examples/cpp/107_index_pyramid.cpp]: This example change has been reverted in a later commit, so the PR no longer modifies `examples/cpp/107_index_pyramid.cpp`.

File: src/algorithm/hgraph_parameter_test.cpp [hunk 4]
```diff
@@ -125,37 +126,58 @@ generate_hgraph_param(const HGraphDefaultParam& param) {
...
+TEST_COMPATIBILITY_CASE("different use reorder", use_reorder, true, false, false)
+TEST_COMPATIBILITY_CASE(
+    "different precise codes io type", precise_codes_io_type, "memory_io", "block_memory_io", true)
+TEST_COMPATIBILITY_CASE("different precise codes quantization type",
+                        precise_codes_quantization_type,
+                        "fp32",
+                        "sq8",
+                        false)
+TEST_COMPATIBILITY_CASE("different use attribute filter", use_attribute_filter, true, false, false)
+TEST_COMPATIBILITY_CASE("different support duplicate", support_duplicate, true, false, false)
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/algorithm/hgraph_parameter_test.cpp]: Reformatted the compatibility test block and re-ran `make fmt` in a follow-up commit.

File: src/algorithm/pyramid.cpp [hunk 5]
```diff
@@ -517,6 +526,7 @@ Pyramid::Add(const DatasetPtr& base, AddMode mode) {
                 no_build_level_index++;
                 continue;
             }
+            ++add_count;
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/algorithm/pyramid.cpp]: Removed the matching `++add_count` increment together with the dead variable.

File: tests/test_pyramid.cpp [hunk 6]
```diff
@@ -106,6 +109,68 @@ PyramidTestIndex::GeneratePyramidSearchParametersString(int64_t ef_search, doubl
...
+    }
+    return ids;
+}
+
+void
+RequireDistancesNearZero(const vsag::DatasetPtr& result, const std::set<int64_t>& expected_ids) {
+    for (int64_t i = 0; i < result->GetDim(); ++i) {
+        if (expected_ids.count(result->GetIds()[i]) != 0) {
+            REQUIRE(std::abs(result->GetDistances()[i]) <= 1e-6F);
+        }
```
[the remark this answers is not in the document]
Comment (inabao) [reply tests/test_pyramid.cpp:170]: Added an explicit `<cmath>` include in `tests/test_pyramid.cpp` so the `std::abs` call no longer relies on transitive includes.

File: src/datacell/sparse_duplicate_tracker.h [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+    Serialize(StreamWriter& writer) const override;
+
+    void
+    Deserialize(StreamReader& reader) override;
+
+    void
+    DeserializeFromLegacyFormat(StreamReader& reader, size_t total_size) override;
+
+    void
+    Resize(InnerIdType new_size) override {
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/datacell/sparse_duplicate_tracker.h:48]: Updated the inline no-op override to omit the unused parameter name.

File: src/algorithm/pyramid.cpp (hunk 2 above)
[the remark this answers is not in the document]
Comment (inabao) [reply src/algorithm/pyramid.cpp]: Fixed in a follow-up commit by removing the unused variable and increment.

File: src/impl/label_table.cpp [hunk 8]
```diff
@@ -40,8 +40,6 @@ LabelTable::LabelTable(Allocator* allocator, bool use_reverse_map, bool compress
       label_remap_(0, allocator),
       use_reverse_map_(use_reverse_map),
       deleted_ids_(allocator),
-      compress_duplicate_data_(compress_redundant_data),
-      duplicate_ids_(0, allocator),
       hole_list_(0, allocator) {
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/impl/label_table.cpp:43]: Fixed in 2e80fd72. The constructor now explicitly ignores `compress_redundant_data` with `(void)compress_redundant_data;` to keep warning-enabled builds clean.

File: src/impl/label_table.cpp [hunk 9]
```diff
@@ -146,6 +144,27 @@ LabelTable::MarkRemove(const std::vector<LabelType>& labels) {
...
+        duplicate_tracker_->DeserializeFromLegacyFormat(reader, label_table_.size());
+    }
+    is_legacy_duplicate_format_ = false;
+
+    if (support_tombstone_) {
+        StreamReader::ReadObj(reader, deleted_ids_);
+    }
+
+    this->total_count_.store(label_table_.size());
+}
```
[the remark this answers is not in the document]
Comment (inabao) [reply src/impl/label_table.cpp:167]: Fixed in 2e80fd72. Added `LabelTable deserializes legacy duplicate payload` in `src/impl/label_table_test.cpp`, which builds a legacy-formatted stream, deserializes it through `LabelTable`, and verifies the attached `DenseDuplicateTracker` is populated as expected.

**Review by LHT129 — APPROVED** [review]
LHTM
